server/s3_post_complete_action.cc

```diff
@@ -424,77 +440,152 @@ bool S3PostCompleteAction::validate_parts() {
+        // set_abort_multipart(true);
+        send_response_to_s3_client();
+        return false;
       }
       object_size += part_metadata->get_content_length();
       awsetag.add_part_etag(part_metadata->get_md5());
       // Remove the entry from parts map, so that in next
       // validate_parts() we dont have to scan it again
       part_kv = parts.erase(part_kv);
+      // Add part object to new object's extended metadata potion
+      add_part_object_to_object_extended(part_metadata, new_object_metadata);
     }
   }
   s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
   return true;
 }
 
-void S3PostCompleteAction::add_object_oid_to_probable_dead_oid_list() {
+void S3PostCompleteAction::add_part_object_to_object_extended(
+    const std::shared_ptr<S3PartMetadata>& part_metadata,
+    std::shared_ptr<S3ObjectMetadata>& object_metadata) {
+  s3_log(S3_LOG_DEBUG, stripped_request_id, "%s Entry\n", __func__);
+  if (!object_metadata) {
+    // Create new object metadata
+    object_metadata = object_metadata_factory->create_object_metadata_obj(
+        request, bucket_metadata->get_object_list_index_oid());
+    object_metadata->set_objects_version_list_index_oid(
+        bucket_metadata->get_objects_version_list_index_oid());
+    // Dummy oid and layout id for new object
+    object_metadata->set_oid(new_object_oid);
+    object_metadata->set_layout_id(layout_id);
+
+    // Generate version id for the new obj as it will become live to s3 clients.
+    object_metadata->regenerate_version_id();
+    // Create extended metadata object and add object parts to it.
+    const std::shared_ptr<S3ObjectExtendedMetadata>& ext_object_metadata =
+        object_metadata->get_extended_object_metadata();
+    if (!ext_object_metadata) {
+      std::shared_ptr<S3ObjectExtendedMetadata> new_ext_object_metadata =
+          object_metadata_factory->create_object_ext_metadata_obj(
+              request, request->get_bucket_name(), request->get_object_name(),
+              object_metadata->get_obj_version_key(), 0, 0,
+              bucket_metadata->get_extended_metadata_index_oid());
+
+      object_metadata->set_extended_object_metadata(new_ext_object_metadata);
+      s3_log(S3_LOG_DEBUG, stripped_request_id,
+             "Created extended object metadata to store object parts");
+    }
+  }
+  const std::shared_ptr<S3ObjectExtendedMetadata>& new_object_ext_metadata =
+      object_metadata->get_extended_object_metadata();
+
+  if (new_object_ext_metadata) {
+    // Add object part to new object's extended metadata
+    struct s3_part_frag_context part_frag_ctx = {};
+    part_frag_ctx.motr_OID = part_metadata->get_oid();
+    part_frag_ctx.PVID = part_metadata->get_pvid();
+    part_frag_ctx.versionID = object_metadata->get_obj_version_key();
+    part_frag_ctx.item_size = part_metadata->get_content_length();
+    part_frag_ctx.layout_id = part_metadata->get_layout_id();
+    part_frag_ctx.is_multipart = true;
+    const int part_number =
```

Comment:
Is there narrow cast?
A compiler or a static analizer can give warning

Reply:
ok. will use correct type

Reply:
Compiler did not shout here...not sure why?